### PR TITLE
Remove opencontainer dependency for timestamping code

### PR DESCRIPTION
**Pull request by priteshbandi**
* timestamping shouldn't depend on opencontainers.
* Also, rephrased some of the error message for certificate chain validation.

### CODE REVIEW

**Review by jondonas — APPROVED** [review]

**Review by rgnote — APPROVED** [review]
LGTM

File: timestamp/request.go [hunk 1]
```diff
@@ -85,15 +88,34 @@ func (r *Request) UnmarshalBinary(data []byte) error {
...
+func validate(digest []byte, alg crypto.Hash) error {
+	l := len(digest)
+	var validContent bool
+	switch alg {
+	case crypto.SHA256:
+		validContent = l == 256/8
+	case crypto.SHA384:
+		validContent =  l == 384/8
+	case crypto.SHA512:
+		validContent = l == 384/8
```
Comment (gokarnm) [inline timestamp/request.go:113]: We are missing unit tests.

```suggestion
	case crypto.SHA512:
		validContent = l == 512/8
```

File: timestamp/http_test.go [hunk 2]
```diff
@@ -75,7 +75,7 @@ func TestHTTPTimestampGranted(t *testing.T) {
 		t.Fatalf("NewHTTPTimestamper() error = %v", err)
 	}
 	message := []byte("notation")
-	req, err := NewRequestFromBytes(message)
+	req, err := NewRequestWithData(message, crypto.SHA256)
 	if err != nil {
 		t.Fatalf("NewRequestFromString() error = %v", err)
```
Comment (gokarnm) [inline timestamp/http_test.go:80]: Method name on line 80 does not match line 78. Same for multiple occurrences in this file.

File: timestamp/request.go [hunk 3]
```diff
@@ -37,36 +38,38 @@ type Request struct {
...
 		CertReq: true,
 	}, nil
 }
 
-// NewRequestFromBytes creates a request based on the given byte slice.
-func NewRequestFromBytes(content []byte) (*Request, error) {
-	return NewRequest(digest.FromBytes(content))
-}
+// NewRequestWithData creates a request based on the given data and hash algorithm.
+func NewRequestWithData(content []byte, alg crypto.Hash) (*Request, error) {
```
Comment (gokarnm) [inline timestamp/request.go:66]: Suggestion - NewRequestWithContent ?

File: timestamp/http_test.go (hunk 2 above)
Comment (gokarnm) [reply timestamp/http_test.go:80]: NewRequestFromString is removed but multiple error messages reference it.

File: timestamp/request.go [hunk 4]
```diff
@@ -37,36 +38,38 @@ type Request struct {
 	Extensions     []pkix.Extension      `asn1:"optional,tag:0"`
 }
 
-// NewRequest creates a request based on the given digest.
-func NewRequest(contentDigest digest.Digest) (*Request, error) {
-	hashAlgorithm, err := getOIDFromDigestAlgorithm(contentDigest.Algorithm())
+// NewRequest creates a request based on the given digest and hash algorithm.
+func NewRequest(digest []byte, alg crypto.Hash) (*Request, error) {
```
Comment (gokarnm) [inline timestamp/request.go:42]: Is there a reason we need both NewRequest/NewRequestWithData , Verify/VerifyWithData ?

**Review by gokarnm — CHANGES_REQUESTED** [review]

File: timestamp/errors.go [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package timestamp
+
+// MalformedRequestError is used when timestamping request is malformed.
+type MalformedRequestError struct {
```
Comment (shizhMSFT) [inline timestamp/errors.go:4]: It is better to name it `ErrorMalformedRequest`. It's a convention to put `Error` as a prefix.

File: timestamp/request.go [hunk 6]
```diff
@@ -1,14 +1,15 @@
 package timestamp
 
 import (
+	"crypto"
 	"crypto/x509/pkix"
 	"encoding/asn1"
-	"encoding/hex"
 	"errors"
+	"fmt"
+	"github.com/notaryproject/notation-core-go/internal/crypto/hashutil"
```
Comment (shizhMSFT) [inline timestamp/request.go:9]: This line should not be with the block of import for built-in packages.

File: timestamp/token.go [hunk 7]
```diff
@@ -7,11 +7,11 @@ import (
 	"encoding/asn1"
 	"errors"
 	"fmt"
+	"github.com/notaryproject/notation-core-go/internal/crypto/hashutil"
```
Comment (shizhMSFT) [inline timestamp/token.go:10]: same here

File: timestamp/token.go [hunk 8]
```diff
@@ -106,8 +106,8 @@ type TSTInfo struct {
 	Extensions     []pkix.Extension `asn1:"optional,tag:1"`
 }
 
-// Verify verifies the message against the timestamp token information.
-func (tst *TSTInfo) Verify(message []byte) error {
+// VerifyWithData verifies the message against the timestamp token information.
+func (tst *TSTInfo) VerifyWithData(message []byte) error {
```
Comment (shizhMSFT) [inline timestamp/token.go:110]: Will `VerifyMessage` be a better name?

File: timestamp/request.go [hunk 9]
```diff
@@ -85,15 +88,34 @@ func (r *Request) UnmarshalBinary(data []byte) error {
...
+	switch alg {
+	case crypto.SHA256:
+		validContent = l == 256/8
+	case crypto.SHA384:
+		validContent =  l == 384/8
+	case crypto.SHA512:
+		validContent = l == 384/8
+	default:
+		return MalformedRequestError{msg: fmt.Sprintf("unsupported hashing algorithm: %s", alg)}
+	}
```
Comment (shizhMSFT) [inline timestamp/request.go:116]: It is better to have `alg.Size()` than hard coded length. 

See https://pkg.go.dev/crypto#Hash.Size

Comment (priteshbandi) [reply timestamp/request.go:116]: Done

File: timestamp/token.go (hunk 8 above)
Comment (priteshbandi) [reply timestamp/token.go:110]: Renamed it to VerifyWithContent.

File: timestamp/request.go [hunk 10]
```diff
@@ -37,36 +38,37 @@ type Request struct {
...
 		CertReq: true,
 	}, nil
 }
 
-// NewRequestFromBytes creates a request based on the given byte slice.
-func NewRequestFromBytes(content []byte) (*Request, error) {
-	return NewRequest(digest.FromBytes(content))
-}
+// NewRequestWithContent creates a request based on the given data and hash algorithm.
+func NewRequestWithContent(content []byte, alg crypto.Hash) (*Request, error) {
```
Comment (shizhMSFT) [inline timestamp/request.go:65]: nit:
```suggestion
func NewRequestFromContent(content []byte, alg crypto.Hash) (*Request, error) {
```

File: timestamp/token.go [hunk 11]
```diff
@@ -106,8 +106,8 @@ type TSTInfo struct {
 	Extensions     []pkix.Extension `asn1:"optional,tag:1"`
 }
 
-// Verify verifies the message against the timestamp token information.
-func (tst *TSTInfo) Verify(message []byte) error {
+// VerifyWithContent verifies the message against the timestamp token information.
+func (tst *TSTInfo) VerifyWithContent(message []byte) error {
```
Comment (shizhMSFT) [inline timestamp/token.go:110]: nit:
```suggestion
func (tst *TSTInfo) VerifyContent(message []byte) error {
```

File: timestamp/request.go [hunk 12]
```diff
@@ -85,15 +87,34 @@ func (r *Request) UnmarshalBinary(data []byte) error {
...
+	case crypto.SHA512:
+		validContent = l == crypto.SHA512.Size()
+	default:
+		return MalformedRequestError{msg: fmt.Sprintf("unsupported hashing algorithm: %s", alg)}
+	}
+	if !validContent {
+		return MalformedRequestError{msg: fmt.Sprintf("digest is of incorrect size: %d", l)}
+	}
+	return nil
+}
```
Comment (shizhMSFT) [inline timestamp/request.go:120]:
```suggestion
func validate(digest []byte, alg crypto.Hash) error {
	switch alg {
	case crypto.SHA256, crypto.SHA384, crypto.SHA512:
		// no-op
	default:
		return MalformedRequestError{msg: fmt.Sprintf("unsupported hashing algorithm: %s", alg)}
	}
	if len(digest) != alg.Size() {
		return MalformedRequestError{msg: fmt.Sprintf("digest is of incorrect size: %d", l)}
	}
	return nil
}
```

**Review by gokarnm — APPROVED** [review]

**Pull request merged**